graph-builder/src/metadata.rs

```diff
@@ -0,0 +1,113 @@
+//! This module implements the fetching of dynamic metadata
+
+use failure::{Error, Fallible};
+use futures::future::Future;
+use quay;
+use registry;
+use std::path::PathBuf;
+use std::rc::Rc;
+
+pub static DEFAULT_QUAY_LABEL_FILTER: &str = "com.openshift.upgrades.graph";
+pub static MANIFESTREF_KEY: &str = "com.openshift.upgrades.graph.release.manifestref";
+
+pub struct MetadataFetchArgs {
+    manifestref: String,
+}
+
+impl MetadataFetchArgs {
+    pub fn new(manifestref: String) -> Self {
+        MetadataFetchArgs { manifestref }
+    }
+}
+
+pub trait AsyncMetadataFetcher<T, E> {
+    fn fetch_metadata(&self, fetch_args: &MetadataFetchArgs) -> Box<Future<Item = T, Error = E>>;
+}
+
+/// Convenience type for a metadata entry
+type MetadataEntry = (String, String);
+
+/// Convenience type for fetch_args()'s metadata_fetcher argument
+pub type MetadataFetcher<'a> = Rc<AsyncMetadataFetcher<Vec<MetadataEntry>, Error> + 'a>;
```

Comment:
@lucab I'm curious about your thoughts on `Rc` vs `Box` here

Reply:
self answer: Rc is not needed here as these instances aren't meant to be owned by multiple sites